### PR TITLE
raftstore: separate read write ready (#10592)

### LINKED ISSUE #10475: Separate read/write ready
### Is your feature request related to a problem? Please describe:
Let's define read ready as ready that won't modify write batch, write ready as ready that modifies write batch. When read/write ready are fetched into one batch, read ready will wait for write ready to finish.

Considering following situation, a full ranges scan will notify all idle regions, which will just propose a read index request to extend read lease. There is no need to do any writes. Internal network latency is usually hundreds of us, but write latency can vary from several to hundreds of milliseconds. So leader can take longer than expected to process a read and cause high tail latency.

### Describe the feature you'd like:
We should separate read/write readiness. #10289 solve the problem by using a separate thread to write raft logs and states. But the change may not be back ported to 4.0/5.0 as changes are large and complicated. For old codebase, we can still optimize the tail latency by adding a new method `light_end` to batch system. `light_end` will be used to release fsms that produce read only readiness. So the procedure of handling ready becomes:
```
begin
fetch fsms
handle normal
    fetch ready
light end
    post_ready for read ready
release read fsms
end
   write disk
   post_ready for write ready
release write fsms
```

I wrote a prototype and the result is promising:

image

The left part is with optimization patch, the right part is v5.0.2. My patch can be see [here](https://github.com/BusyJay/tikv/commit/4b65f35f64f94c500ddf36336871c258ef447f8c)

### Describe alternatives you've considered:
Do nothing for other branch and wait for #10289.

But the change in my prototype is quite small, I think it's better accept the change and optimize for all versions.

### PR DESCRIPTION
**Pull request by BusyJay**
### What problem does this PR solve?

Close #10475.

Read ready can be processed early to reduce read latency.

### Check List 

Tests 

- Unit test
- Integration test
- Manual test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
- separate read write ready to reduce read latency
```

### CODE REVIEW

File: components/raftstore/src/store/fsm/store.rs [hunk 1]
```diff
@@ -877,27 +905,46 @@ impl<EK: KvEngine, ER: RaftEngine, T: Transport> PollHandler<PeerFsm<EK, ER>, St
...
+            peer.update_memory_trace(&mut self.trace_event);
+        }
+        MEMTRACE_PEERS.trace(mem::take(&mut self.trace_event));
         if self.poll_ctx.has_ready {
-            self.handle_raft_ready(peers);
+            self.handle_raft_ready(peers, to_skip_end);
+        }
+    }
+
+    fn end(&mut self, peers: &mut [Option<impl TrackedFsm<Target = PeerFsm<EK, ER>>>]) {
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/store.rs:945]: @5kbpers PTAL, I'm not sure if end is the correct place for pending latency inspector after introducing `light_end`.

**Review by gengliqi — APPROVED** [review]

File: components/raft_log_engine/src/engine.rs [hunk 2]
```diff
@@ -74,8 +74,12 @@ impl RaftLogBatchTrait for RaftLogBatch {
         Ok(())
     }
 
+    #[inline]
     fn persist_size(&self) -> usize {
-        panic!("persist_size is not implemented for raft engine");
+        // TODO: This is not data size, but sufficient for current usage.
```
Comment (BusyJay) [inline components/raft_log_engine/src/engine.rs:79]: /cc @tabokie as a reminder.

**Review by NingLin-P — APPROVED** [review]

File: components/raftstore/src/store/fsm/store.rs (hunk 1 above)
Comment (5kbpers) [reply components/raftstore/src/store/fsm/store.rs:945]: The purpose of inspectors is to inspect write latency so I think it's ok.

File: components/raft_log_engine/src/engine.rs (hunk 2 above)
Comment (tabokie) [reply components/raft_log_engine/src/engine.rs:79]: It was implemented in https://github.com/tikv/raft-engine/pull/82 a few days ago.

**Review by NingLin-P — APPROVED** [review]

**Pull request merged**